### PR TITLE
Design Preview: Fix infinite re-rendering

### PR DESCRIPTION
**Pull request by arthur791004**
Related to p1697776363743409-slack-CRWCHQGUB

## Proposed Changes

* Avoid calling the `receiveCurrentUser` action on every render
* Avoid initializing the preselected design and style variations multiple times because the deps change, e.g. `pickUnlistedDesign`. The `pickUnlistedDesign` function is always the new one on every render

## Testing Instructions

* Go to /setup/site-setup/designSetup?theme=fewer&siteSlug=your_site
* Ensure the page won't re-rendering forever
* Ensure the back button works as expected

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by taipeicoder** [comment]
>  The `pickUnlistedDesign` function is always the new one on every render.

This sounds like a potential source of more bugs in the future. Could we add a FIXME or TODO in the comment to warn this?

File: client/landing/stepper/index.tsx [hunk 1]
```diff
@@ -51,26 +59,22 @@ function determineFlow() {
...
-	user,
-	flow,
-} ) => {
-	const { receiveCurrentUser } = useDispatch( USER_STORE );
+const FlowSwitch: React.FC< { flow: Flow } > = ( { flow } ) => {
 	const { setEcommerceFlowRecurType } = useDispatch( ONBOARD_STORE );
+	const recurType = useQuery().get( 'recur' ) ?? '';
 
-	const recurType = useQuery().get( 'recur' );
+	useEffect( () => {
```
Comment (taipeicoder) [inline client/landing/stepper/index.tsx]: Let's keep Vertex in the loop about this change @Automattic/vertex

**Review by taipeicoder — APPROVED** [review]
Nice fix 🎉

**Comment by arthur791004** [comment]
> Could we add a FIXME or TODO in the comment to warn this?

We don't need any changes for this as we have avoided setting the state when any dependency changes in the future. We can wrap the function with the `useCallback` to prevent it from becoming a new one on every render but people might add a new function to the deps of this effect. Hence, using a new variable to record whether it's initialized has completely solved this issue.

**Comment by fushar** [comment]
This looks like an anti-pattern in React's `useEffect()` usage... give me a few minutes to understand and look into it a bit more 😄

**Review by fushar — CHANGES_REQUESTED** [review]
I feel like patching `useRef` to a `useEffect` is just hiding the bug rather than fixing it.

The issue is that we're adding `pickUnlistedDesign` dependency to an effect that has a job to set the design from the URL, which will always set the state (via `setSelectedDesign`) and will rerender. However, `pickUnlistedDesign` is a function which changes reference every rerender, and thus the infinite loop.

As I was the one who added the `pickUnlistedDesign` dependency in the codebase, I feel obligated to correct the issue 🙈 

I am proposing this, which fixes the root cause. What do you think?

```diff
diff --git a/client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/hooks/use-recipe.ts b/client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/hooks/use-recipe.ts
index d140e98b71..69c32e3cae 100644
--- a/client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/hooks/use-recipe.ts
+++ b/client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/hooks/use-recipe.ts
@@ -188,7 +188,7 @@ const useRecipe = (
 				: design.slug === preselectedTheme
 		);
 		if ( ! requestedDesign ) {
-			pickUnlistedDesign( preselectedTheme );
+			setSelectedDesign( undefined );
 			return;
 		}
 
@@ -206,11 +206,16 @@ const useRecipe = (
 		preselectedTheme,
 		preselectedStyle,
 		allDesigns,
-		pickUnlistedDesign,
 		setSelectedDesign,
 		setSelectedStyleVariation,
 	] );
 
+	useEffect( () => {
+		if ( preselectedTheme && ! selectedDesign ) {
+			pickUnlistedDesign( preselectedTheme );
+		}
+	}, [ preselectedTheme, selectedDesign, pickUnlistedDesign ] );
+
 	/**
 	 * Initialize the preselected colors
 	 */
```

**Comment by arthur791004** [comment]
> I feel like patching useRef to a useEffect is just hiding the bug rather than fixing it.

I don't think so. Sometimes we still need a flag to avoid doing a thing multiple times.

> I am proposing this, which fixes the root cause. What do you think?
> ```js
>	useEffect( () => {
>		if ( preselectedTheme && ! selectedDesign ) {
>			pickUnlistedDesign( preselectedTheme );
>		}
>	}, [ preselectedTheme, selectedDesign, pickUnlistedDesign ] );
> ```

This seems to be buggy.
* The `! selectedDesign` is always true before we fetch the `allDesigns`, so people will be redirected to the Theme Showcase when they land on the Design Picker with the preselected design directly.
* If people land on the Design Picker with the `preselectedTheme`, they will be redirected as well when they click the Back button because we will unset the `selectedDesign` at that time.

But I'm fine with both ideas 🙂

**Comment by arthur791004** [comment]
@fushar I made change, https://github.com/Automattic/wp-calypso/pull/83286/commits/cc8f071668dedcb062553be007c1e6e0860dcae6, as you suggested 🙂

**Review by fushar — APPROVED** [review]
> I don't think so. Sometimes we still need a flag to avoid doing a thing multiple times.

Yes but I think `useRef()` should be used as a last resort when we can't control things in React. I would even argue that `useCallback()` is better here if it's still possible.

> This seems to be buggy.
> @fushar I made change, https://github.com/Automattic/wp-calypso/commit/cc8f071668dedcb062553be007c1e6e0860dcae6, as you suggested

Thanks!! Just realized that mine is buggy :') and your version is better! Tested and it works (both for normal case and picking unlisted design) 👍 🚢

**Comment by fushar** [comment]
Suggestion: do we need to split this PR so that the one that fixes `receiveCurrentUser()` is in another PR?

I think the one that we fix in `useRecipe()` should be shipped ASAP.

**Comment by arthur791004** [comment]
> Suggestion: do we need to split this PR so that the one that fixes receiveCurrentUser() is in another PR?

Ok even though I think it should work as expected 🙈

**Review by fushar — APPROVED** [review]

**Pull request merged**